06-TimeManagement/itime.c

```diff
@@ -0,0 +1,174 @@
+#include <linux/init.h>
+#include <linux/module.h>
+#include <linux/device.h>
+#include <linux/err.h>
+#include <linux/time.h>
+#include <linux/jiffies.h>
+#include <linux/timer.h>
+
+static struct timer_list timer_hendler;
+static u64 timerperiod_ms = 1000;
+
+static void timer_cb(struct timer_list *l)
+{
+	static u32 ncb;
+
+	pr_info("itime: %s: %u\n", __func__, ncb++);
+	mod_timer(&timer_hendler, jiffies + msecs_to_jiffies(timerperiod_ms));
+}
+
+static void timerinit(void)
+{
+	timer_setup(&timer_hendler, timer_cb, 0);
+	mod_timer(&timer_hendler, jiffies + msecs_to_jiffies(timerperiod_ms));
+}
+
+static void timerdeinit(void)
+{
+	del_timer(&timer_hendler);
+}
+
+/**
+ * return: jiffies diff
+ */
+static u64 getdiff(void)
+{
+	static u64 jiffiescach;
+	u64 jiffiesdiff;
+
+	jiffies = get_jiffies_64();
+	jiffiesdiff = jiffies - jiffiescach;
+	jiffiescach = jiffies;
+	return jiffiesdiff;
+}
+
+/**
+ * sysfs callbacks
+ */
+static ssize_t show_rtime(struct class *class, struct class_attribute *attr,
+						char *pbuf)
+{
+	size_t strlen;
+
+	pr_info("itime: %s\n", __func__);
+	strlen = sprintf(pbuf, "%08llu\n", getdiff() / HZ);
+
+	return strlen + 1;
+}
+
+static ssize_t show_atime(struct class *class, struct class_attribute *attr,
+						char *pbuf)
+{
+	size_t strlen;
+	struct timeval tv;
+	struct tm tm;
+
+	pr_info("itime: %s\n", __func__);
+	jiffies_to_timeval(getdiff(), &tv);
+	time_to_tm(tv.tv_sec, 0, &tm);
+	strlen = sprintf(pbuf, "%02i:%02i:%02i\n", tm.tm_hour, tm.tm_min,
+					tm.tm_sec);
+
+	return strlen + 1;
+}
+
+static ssize_t store_adjtime(struct class *class,
+							struct class_attribute *attr,
+							const char *pbuf, size_t count)
+{
+	int n, ms;
+
+	n = sscanf(pbuf, "%u", &ms);
+	if (n < 1) {
+		pr_info("itime: %s: error time\n", __func__);
+		return count;
+	}
+
+	if (ms < 100)
+		ms = 100;
+	if (ms > 10000)
+		ms = 10000;
```

Comment:
Is there any way to stop the timer?

Reply:
?
Its limiter.

Reply:
No. add?

Reply:
Yes, please.
Since without this feature this timer will overflood the log.

Reply:
Done.